src/socket_base.cpp

```diff
@@ -1289,6 +1295,7 @@ int zmq::socket_base_t::close ()
     //  process.
     send_reap (this);
 
+    EXIT_MUTEX ();
```

Comment:
Adding sync.lock/unlock to the destructor doesn't seem to make a difference. I'll send a PR to move the unlock earlier.